### ISSUE TITLE
New metrics and different Loss in TF version of Segformer

### ISSUE THREAD

**Issue by ebgoldstein** [huggingface/transformers#22092]
### Feature request

I am playing with the awesome [Segformer finetuning example on the Keras website](https://keras.io/examples/vision/segformer/) made by @sayakpaul that relies on HF Transformers. In this example no loss function nor any metrics are specified in  `model.compile()` . I would like to be able to add metrics (e.g., IoU, Dice, etc), and potentially change the loss  for the Segformer model.

When I tried to make these additions,  the compile step failed.  (From reading the Segformer paper and original code it seems like all metrics and losses need to have some form of masking?).  

Any advice or info on how to implement these changes would be awesome (and i apologize in advance if I have missed the relevant docs (i did look!). 

(based on comms with @sayakpaul , i am also cc:ing @Rocketknight1 )

### Motivation

Track various metrics during the fine-tuning of the Segformer model.

### Your contribution

I think once i understand the solution steps i would be able to determine if i could contribute

**Comment by Rocketknight1** [comment]
Interesting! Can you document exactly what error you got with the compile step and what code you ran to cause them?

**Comment by ebgoldstein** [comment]
Hi @Rocketknight1 
I misremembered- the error is not after `model.compile()` - compiling a model with a different loss function, added metrics, a custom loss, or custom metrics all compile w/ no error. The errors appear with `model.fit()` .

So far I have tried to fit a model with a range of things :
- a custom loss (my own version of Dice Loss)
- added metrics (`tf.keras.metrics.MeanIoU()` and/or a (custom) Dice metric)
- using KLDivergence loss (`tf.keras.losses.KLDivergence()`)

All produce errors during `model.fit()`, and all produce their own sets of errors.. All of them seem to me to be some type of tensorshape issue, but the Tracebacks are all different. To make sure its not just me, my colleague @dbuscombe-usgs has also tried, and also reported similar issues (with different datasets, different number of classes, different TF versions,  different machines, etc.). I can provide a reference dataset and the scripts I am working with, if needed...

**Comment by Rocketknight1** [comment]
Yes please! Ideally if you could give us some minimal code that reproduces the issue, that would make it much easier for us to track it down.

Also, sorry for the delay in replying here - I was away on St. Patrick's Day so I'm only getting to my GitHub backlog now!

**Comment by Rocketknight1** [comment]
Ah, I see! The issue here is caused by some specific behaviour of the SegFormer models when using inputs of this resolution. The model outputs are actually at a lower resolution than the inputs - you can check this by manually passing in a batch. The output logits come out at 128x128, whereas the input is 512x512. This results in the loss computation failing because the logit and label tensors can't be aligned with each other.

If you use the model's [internal loss computation](https://github.com/huggingface/transformers/blob/v4.27.2/src/transformers/models/segformer/modeling_tf_segformer.py#L793-L811) by not passing any loss argument to `compile()`, then logits are upscaled before applying the cross-entropy loss and training works correctly. If you want to use your own custom loss function you'll have to do something similar.

I'm not sure exactly why the output resolution for SegFormer is different from the input resolution, but it's not a bug in the Hugging Face TensorFlow implementation because the original model and our PyTorch implementation do this as well. @sayakpaul do you know why the model does that?

**Comment by ebgoldstein** [comment]
thx for that code highlight @Rocketknight1 , super helpful and i understand it now -- i would need a similar upsampling routine.

related Q to finding the output resolution - is there a reason that `summary()` does not provide info on all the layers/internal architecture of the model?

**Comment by Rocketknight1** [comment]
Yeah, refactoring our TF models to make `summary()` more usable is absolutely on the list! Unfortunately it's quite a big list, but it's definitely there.

**Comment by ebgoldstein** [comment]
Awesome, thanks so much for all the helpful info @Rocketknight1 & @sayakpaul . I can close this issue now as i understand the landscape much better and it seems the requested feature is already on your list!
thanks again - i really appreciate it!
